render/storagerenderer.go

```diff
@@ -146,3 +148,57 @@ func (b blockDeviceClaimToBlockDeviceRenderer) Render(ctx context.Context, rpt r
 	}
 	return Nodes{Nodes: nodes}
 }
+
+// CSPCToCSPIRenderer is a Renderer which produces a renderable kubernetes CRD CSPC
+var CSPCToCSPIRenderer = cspcToCSPIRenderer{}
+
+// cspcToCSPIRenderer is a Renderer to render CSPC & CSPI nodes.
+type cspcToCSPIRenderer struct{}
+
+// Render renders the SPC & CSP nodes with adjacency.
+// Here we are obtaining the spc name from csp and adjacency is created by matching it with spc name.
+func (v cspcToCSPIRenderer) Render(ctx context.Context, rpt report.Report) Nodes {
+	nodes := make(report.Nodes)
+	for cspcID, cspcNode := range rpt.CStorPoolCluster.Nodes {
+		cspcName, _ := cspcNode.Latest.Lookup(kubernetes.Name)
+		cspcNamespace, _ := cspcNode.Latest.Lookup(kubernetes.Namespace)
+		for cspiID, cspiNode := range rpt.CStorPoolInstance.Nodes {
+			spcName, _ := cspiNode.Latest.Lookup(kubernetes.StoragePoolClaimName)
+			cspiNamespace, _ := cspiNode.Latest.Lookup(kubernetes.Namespace)
+			if cspcName == spcName && cspcNamespace == cspiNamespace {
+				cspcNode.Adjacency = cspcNode.Adjacency.Add(cspiID)
+				cspcNode.Children = cspcNode.Children.Add(cspiNode)
+			}
+		}
+		nodes[cspcID] = cspcNode
+	}
+	return Nodes{Nodes: nodes}
+}
+
+// CSPIToBDRenderer is a renderer which produces a renderable CRD CSPI.
+var CSPIToBDRenderer = cspiToBDRenderer{}
+
+// cspiToBDRenderer is a Renderer to render CSPI & BD nodes.
+type cspiToBDRenderer struct{}
+
+func (n cspiToBDRenderer) Render(ctx context.Context, rpt report.Report) Nodes {
+	nodes := make(report.Nodes)
+	for cspiID, cspiNode := range rpt.CStorPoolInstance.Nodes {
+		cspiNamespace, _ := cspiNode.Latest.Lookup(kubernetes.Namespace)
+		blockDeviceList, _ := cspiNode.Latest.Lookup(kubernetes.BlockDeviceList)
+		blockDevices := strings.Split(blockDeviceList, report.ScopeDelim)
+		for _, blockDevice := range blockDevices {
+			for bdID, bdNode := range rpt.BlockDevice.Nodes {
```

Comment:
Please make use of another way we discussed offline, that will reduce one loop here.